### ISSUE TITLE
PHP 8: readErrorMessage fails

### ISSUE THREAD

**Issue by ilya-konanykhin** [move-backend/ApnsPHP#28]
This line fails with TypeError in PHP 8:

```
$errorMessage = @fread($this->hSocket, self::ERROR_RESPONSE_SIZE);
// PHP Fatal error:  Uncaught TypeError: fread(): Argument #1 ($stream) must be of type resource, CurlHandle given
```

Guess this is a common problem with CURL handles in PHP 8, compare: https://github.com/symfony/symfony/issues/37378

Unfortunately, seems like there's no easy fix as the whole method relies on reading from a socket rather than working via CURL, but could be wrong, I'm no expert.

If you give a hint on how to fix this I could try building a PR.

**Comment by ilya-konanykhin** [comment]
Don't know if its worth mentioning this is related to the HTTP protocol only since the binary seems gone for good now.

**Comment by SMillerDev** [comment]
I think we'd have to split the logic into a resource and an object flow. Unless there is a way to do this completely through cURL (which might be cleaner in the long run).

**Comment by ilya-konanykhin** [comment]
> completely through cURL

As far as I got it, the binary protocol is dead. Probably it'd be better to remove its remnants altogether and work through CURL solely, but this requires a massive cleanup...

**Comment by SMillerDev** [comment]
Everything about this project requires massive cleanup 😅

**Comment by jruston** [comment]
Is there any update on how this issue might be addressed? Would be nice to have the library fully working on PHP 8.

**Comment by ilya-konanykhin** [comment]
Sorry, no news on my end.

**Comment by jruston** [comment]
It seems that the fread function shouldn't be called at all when using the HTTP protocol. On PHP 7.4 it shows a warning as well, but this is being hidden by the @. The only difference with PHP 8 is that it is now a fatal error. Refactoring readErrorMessage to avoid fread when using the HTTP protocol should fix it, I would think.

**Comment by SMillerDev** [comment]
We're often a version behind the latest major php release so M2mobi will be affected by this eventually too. But either removing the binary protocol or refactoring the fread flow out of the http flow sounds good to me.

**Comment by pprkut** [comment]
We just released 3.0.0 beta 1, which among other things has full compatibility with PHP 8. Sorry this took so long to fix on our end!
